opal/corelib/array.rb

```diff
@@ -10,6 +10,28 @@ class Array < `Array`
   `Opal.defineProperty(self.$$prototype, '$$is_array', true)`
 
   %x{
+    // Recent versions of V8 (> 7.1) only use an optimized implementation when Array.prototype is unmodified.
+    // For instance, "array-splice.tq" has a "fast path" (ExtractFastJSArray, defined in "src/codegen/code-stub-assembler.cc")
+    // but it's only enabled when "IsPrototypeInitialArrayPrototype()" is true.
+    //
+    // Older versions of V8 were using relatively fast JS-with-extensions code even when Array.prototype is modified:
+    // https://github.com/v8/v8/blob/7.0.1/src/js/array.js#L599-L642
+    //
+    // In short, Array operations are slow in recent versions of V8 when the Array.prototype has been tampered.
+    // So, when possible, we are using faster open-coded version to boost the performance.
+
+    // As of V8 8.4, depending on the size of the array, this is up to ~25x times faster than Array#shift()
+    // Implementation is heavily inspired by: https://github.com/nodejs/node/blob/ba684805b6c0eded76e5cd89ee00328ac7a59365/lib/internal/util.js#L341-L347
+    function shiftNoArg(list) {
+      var index = 0;
+      var r = list[index]
+      for (; index + 1 < list.length; index++) {
+        list[index] = list[index + 1];
```

Comment:
This is just a nit, but if you have time to fix it that'd be great!

Can you please use an explicit `0` for `var r`, like `var r = list[0]`.
Also I think it might be better to memoize the length and trade two `+ 1` for a `- 1`, like:

```js
var index = 1
var length
for (; index < list.length; index++) {
          list[index - 1] = list[index];
}
```

Reply:
Sure :+1:

Reply:
Performance wise I don't see any change... I guess the bytecode produced by V8 is the same (or at least equivalent in term of performance)